### ISSUE TITLE
SVM output layer in tensor flow

### ISSUE THREAD

**Issue by lixiangchun** [tensorflow/tensorflow#9270]
The SVM output layer has been shown to accelerate model training in [https://arxiv.org/pdf/1306.0239.pdf](https://arxiv.org/pdf/1306.0239.pdf).  Is the SVM output layer available as an alternative for `tf.nn.softmax` in tensorflow at this moment?

**Comment by drpngx** [comment]
I'm not aware of an effort here, but that's probably a good candidate for `tf.contrib`. We could put it in learn @sguada , layers @fchollet or `nn` @ebrevdo .

**Comment by asimshankar** [comment]
@agarwal-ashish might also know of any existing efforts

**Comment by petrosmol** [comment]
There is no such a layer currently. We have the following that are related but not sufficient for what you want to do here:
- a binary svm head (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/head.py) which is the head for standard hinge loss
- a linear SVM classifier (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/svm.py). This is for binary classification only and also uses SDCA for the underlying optimization problem but SDCA is not gradient based so it won't play well when composing with SGD and NNs.

However, I was planning to add multi-class hinge loss (Cramer & Singer maybe) at some point.

**Comment by drpngx** [comment]
Thanks! Sounds good.

**Comment by lixiangchun** [comment]
@petrosmol @drpngx  there is already an implementation of [SVM output layer](https://github.com/dmlc/mxnet/pull/2708/commits/7a651fd2867e56ffcbf539aaaf4218e11cf61b0d) in MxNet. It may be useful for similar implementation in tensorflow.

**Comment by ebrevdo** [comment]
If you can write the SVM layer as a composition of simple ops, inputs, and
Variables, i suggest putting it in tf.contrib.layers.  How much C++ code
does it require?

**Comment by isabel-schwende** [comment]
Has there been any progress about this implementation? I'd be also interested in an option to use one-class SVM. I'm currently considering implementing that myself but it might take a while if I have to do it from scratch...

**Comment by petrosmol** [comment]
FYI, I have checked in sparse_multiclass_hinge_loss which can be used (almost) as a drop-in replacement in place of sparse_softmax_cross_entropy_with_logits. See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/losses.py
for more details. The only difference is that logits passed to sparse_multiclass_hinge_loss should have rank 2 ([batch_size, num_classes]). And I ran some experiments that confirm that, in some cases, accuracy-related metrics can be improved by replacing softmax_cross_entropy with multiclass_hinge_loss

**Comment by errord** [comment]
+1

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by atomextranova** [comment]
Is there any further update that supports directly make SVM, especially non-linear svm with custom regularization (at least common ones like l2-regularization) as one layer in a neural network?

**Comment by cybot16** [comment]
Are there any updates on this topic?

**Comment by petrosmol** [comment]
Is this a separate feature request? In response to the initial request I checked in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/losses.py some time ago. 

@atomextranova: Can you elaborate on your request? also shall we open a separate issue for this?

**Comment by cybot16** [comment]
Thanks!
